[addon][tools] fix dll load helper

AlwinEsch opened the pull request:
## Description

The header `<kodi/tools/DllHelper.h>` was now longer not usable, with lack of Android dll supported.
Currently have all addons where use them it by hack inside his parts.

This add the support about copy of requested dll for Android.

Further is the documentation reworked to have shown in doxygen.

As another part where a bit harder, is that as addon depends the tools are added. Before was the header not checked about changes where makes it harder in case of changes.

This then on addons build by use:
```
-- Addon dependency check ...
--  - Added API usage definition: ADDON_GLOBAL_VERSION_MAIN_USED (Version: "1.2.0", Min. Version: "1.2.0")
--  - Added API usage definition: ADDON_GLOBAL_VERSION_FILESYSTEM_USED (Version: "1.1.0", Min. Version: "1.1.0")
--  - Added API usage definition: ADDON_GLOBAL_VERSION_TOOLS_USED (Version: "1.0.0", Min. Version: "1.0.0")
--  - Added API usage definition: ADDON_INSTANCE_VERSION_AUDIODECODER_USED (Version: "2.0.1", Min. Version: "2.0.1")
```

<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->
As now the other request was there, where needed update of all, thought to make this short after them.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):
Picture of docs:
![Bildschirmfoto vom 2020-05-19 02-34-46](https://user-images.githubusercontent.com/6879739/82272194-2a1fac00-997a-11ea-956f-9a49782ba972.png)

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

AlwinEsch committed:
[addons][tools] bring parts of CDllHelper in namespace kodi::tools

This to have more safe if something else have this name too.

AlwinEsch committed:
[addons][tools] update dll helper to support Android

This was already before on the addons where needed, as long it was not
fixed about Android in Kodi itself.

This add the there needed copy of Dll to anotther place so that it can be
used.

AlwinEsch committed:
[addons][tools] update documentation of CDllHelper

There become sthe needed documentation updated and cleaned up to have
usable and shown inside dev-kit doxygen.

Also is there a group about "tools" added where this is included.

garbear commented on system/addon-manifest.xml:
```diff
@@ -9,6 +9,7 @@
   <addon>kodi.binary.global.network</addon>
   <addon>kodi.binary.global.gui</addon>
   <addon>kodi.binary.global.filesystem</addon>
+  <addon optional="true">kodi.binary.global.tools</addon>
```
Why optional?

AlwinEsch replied:
Thought to prevent problems, if them not as dummy addon in Kodi.
Also part not work direct with Kodi. 
Can also make not optional.

garbear replied:
I would consider making mandatory, unless there's some way or reason to configure kodi without this

AlwinEsch replied:
Have changed.

garbear commented:
I'm surprised that Android can load a DLL that was written by the application at runtime. Do I misunderstand, or is that actually the case?

AlwinEsch commented:
On Android the dll handling is a bit funny created, the addon lib is installed within user path where not allow exectuable parts.

But the copy to Kodi's lib folder works. Also the addon itself does that way, e.g. here https://github.com/xbmc/xbmc/blob/master/xbmc/addons/binary-addons/AddonDll.cpp#L53-L83

AlwinEsch committed:
[addons][tools] add helper as ADDON_GLOBAL_TOOLS to versions.h

This to allow versions check about, before was them not included in
check where makes it dangerous by changes.

AlwinEsch commented:
jenkins build this with addons please

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
jenkins build this please
